From 943652bb41637d0fcfc52aad92247cb0fcf88164 Mon Sep 17 00:00:00 2001
From: Joey Riches <josephriches@gmail.com>
Date: Wed, 11 May 2022 23:55:42 +0100
Subject: [PATCH] Enable gamescope usage on nvidia >= 515

Check we're using 515 or greater major version before allowing
gamescope usage on nvidia.

515.43.04 was the first driver to support VK_EXT_image_drm_format_modifier,
which is required by gamescope to function.
---
 lutris/sysoptions.py |  3 +--
 lutris/util/linux.py | 17 +++++++++++++++++
 2 files changed, 18 insertions(+), 2 deletions(-)

diff --git a/lutris/sysoptions.py b/lutris/sysoptions.py
index 688ac4ff07..898aca02bf 100644
--- a/lutris/sysoptions.py
+++ b/lutris/sysoptions.py
@@ -8,7 +8,6 @@
 from lutris import runners
 from lutris.util import linux, system
 from lutris.util.display import DISPLAY_MANAGER, SCREEN_SAVER_INHIBITOR, USE_DRI_PRIME
-from lutris.util.graphics import drivers
 
 VULKAN_DATA_DIRS = [
     "/usr/local/etc/vulkan",  # standard site-local location
@@ -184,7 +183,7 @@ def get_vk_icd_choices():
         "label": _("Enable gamescope"),
         "default": False,
         "advanced": True,
-        "condition": bool(system.find_executable("gamescope")) and not drivers.is_nvidia(),
+        "condition": bool(system.find_executable("gamescope")) and linux.LINUX_SYSTEM.nvidia_gamescope_support(),
         "help": _("Use gamescope to draw the game window isolated from your desktop.\n"
                   "Use Ctrl+Super+F to toggle fullscreen"),
     },
diff --git a/lutris/util/linux.py b/lutris/util/linux.py
index 2abbae1761..e986730764 100644
--- a/lutris/util/linux.py
+++ b/lutris/util/linux.py
@@ -215,6 +215,23 @@ def gamemode_available(self):
             return True
         return False
 
+    def nvidia_gamescope_support(self):
+        """ Return whether gamescope is supported if we're on nvidia"""
+        if not drivers.is_nvidia():
+            return True
+
+        # 515.43.04 was the first driver to support
+        # VK_EXT_image_drm_format_modifier, required by gamescope.
+        minimum_nvidia_version_supported = 515
+        driver_info = drivers.get_nvidia_driver_info()
+        driver_version = driver_info["nvrm"]["version"]
+        major_version = int(driver_version.split(".")[0])
+
+        if not major_version >= minimum_nvidia_version_supported:
+            return False
+        else:
+            return True
+
     @property
     def has_steam(self):
         """Return whether Steam is installed locally"""
